src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts

```diff
@@ -0,0 +1,427 @@
+  )
+  .fn(t => {
+    const { indexed, indirectUsage, useDrawCountBuffer, drawCountUsage } = t.params;
+
+    const indirectBuffer = t.createBufferTracked({
+      size: 256,
+      usage: indirectUsage,
+    });
+    const drawCountBuffer = useDrawCountBuffer
+      ? t.createBufferTracked({
+          size: 256,
+          usage: drawCountUsage,
+        })
+      : undefined;
+
+    const { encoder, validateFinish } = t.createEncoder('render pass');
+    encoder.setPipeline(t.createNoOpRenderPipeline());
+    if (indexed) {
+      encoder.setIndexBuffer(t.makeIndexBuffer(), 'uint32');
+      encoder.multiDrawIndexedIndirect(indirectBuffer, 0, 1, drawCountBuffer);
+    } else {
+      encoder.multiDrawIndirect(indirectBuffer, 0, 1, drawCountBuffer);
+    }
+    const shouldSucceed =
+      (indirectUsage & GPUBufferUsage.INDIRECT) !== 0 &&
+      (!drawCountBuffer || drawCountUsage & GPUBufferUsage.INDIRECT) !== 0;
+    validateFinish(shouldSucceed);
+  });
+
+g.test('offsets_alignment')
+  .desc(
+    `
+Tests indirect and draw count offsets must be a multiple of 4.
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+  .paramsSubcasesOnly(
+    kIndirectMultiDrawTestParams
+      .combine('indirectOffset', [0, 2, 4] as const)
+      .combine('drawCountOffset', [0, 2, 4] as const)
+  )
+  .fn(t => {
+    const { indexed, indirectOffset, useDrawCountBuffer, drawCountOffset } = t.params;
+
+    const indirectBuffer = t.createBufferTracked({
+      size: 256,
+      usage: GPUBufferUsage.INDIRECT,
+    });
+    const drawCountBuffer = useDrawCountBuffer
+      ? t.createBufferTracked({
+          size: 256,
+          usage: GPUBufferUsage.INDIRECT,
+        })
+      : undefined;
+
+    const { encoder, validateFinish } = t.createEncoder('render pass');
+    encoder.setPipeline(t.createNoOpRenderPipeline());
+    if (indexed) {
+      encoder.setIndexBuffer(t.makeIndexBuffer(), 'uint32');
+      encoder.multiDrawIndexedIndirect(
+        indirectBuffer,
+        indirectOffset,
+        1,
+        drawCountBuffer,
+        drawCountOffset
+      );
+    } else {
+      encoder.multiDrawIndirect(
+        indirectBuffer,
+        indirectOffset,
+        1,
+        drawCountBuffer,
+        drawCountOffset
+      );
+    }
+
+    // We need to figure out if https://github.com/gpuweb/gpuweb/pull/2315/files#r1773031950 applies.
+    validateFinish(indirectOffset % 4 === 0 && (!useDrawCountBuffer || drawCountOffset % 4 === 0));
+  });
+
+g.test('indirect_offset_oob')
+  .desc(
+    `
+Tests multi indirect draw calls with various offsets and buffer sizes.
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+  .params(u =>
+    u.expandWithParams(() => {
+      return [
+        // In bounds
+        { bufferSize: 4, indirectOffset: 0, _valid: true },
```

Comment:
nit: Could we combine these test cases with the ones below as a variable? If I didn't miss anything these cases seem similar with small differences, and combining them would be cleaner/more readable.

Reply:
+1, if possible store this array as a variable and use it twice. A pattern we use elsewhere to make this work is instead of having `bufferSize` as a parameter, have something like `bufferExtraSize` which has negative, zero, and positive cases (`4 + (-1) = 3` for indirect, `5 + (-1) = 4` for indexed indirect)

And if there are any cases that are specific to one or the other then you can do:

```ts
const kIndirectOffsetOOBCommonCases = [
  /* ... */
] as const;

/* ... */
  .combineWithParams([
    ...kIndirectOffsetOOBCommonCases,
    /* any additional cases which are not common between the two */
  ])
```

Reply:
I've hopefully addressed your feedback in latest patch. Please have a look!

Reply:
Sorry for the late response, but the patch looks good to me!